network/subscriber.go

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2017 The Alvalor Authors
+//
+// This file is part of Alvalor.
+//
+// Alvalor is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Alvalor is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with Alvalor.  If not, see <http://www.gnu.org/licenses/>.
+
+package network
+
+import (
+	"github.com/rs/zerolog"
+	"sync"
+	"time"
+)
+
+func handleSubscriber(log zerolog.Logger, wg *sync.WaitGroup, subscriber chan interface{}, subscribers map[chan<- interface{}][]func(interface{}) bool) {
+	defer wg.Done()
+
+	log = log.With().Str("component", "subscriber").Logger()
+	log.Debug().Msg("subscriber routine started")
+	defer log.Debug().Msg("subscriber routine stopped")
+Loop:
+	for {
+		select {
+		case msg, ok := <-subscriber:
+			if !ok {
+				break Loop
+			}
+			triggerSubscribers(log, msg, subscribers)
+		}
+	}
+}

```

Comment:
I think a better architecture would be to:
1) Have one handler for what you call "inner subscriber", but it will be simply the stream to process internally.
2) Have one handler for each "outer subscriber"; whenever you subscribe to the package, you will launch one of them; whenever you unsubscribe, you stop it.
The difficulty will be to nicely multiplex from what comes from the inner stream to a variable number of outer streams.